cloudcredential/v1/register.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2018 The OpenShift Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1 contains API Schema definitions for the cloudcredential v1 API group
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen=package,register
+// +k8s:conversion-gen=github.com/openshift/api/cloudcredential
+// +k8s:defaulter-gen=TypeMeta
+// +groupName=cloudcredential.openshift.io
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	GroupName     = "cloudcredential.openshift.io"
+	GroupVersion  = schema.GroupVersion{Group: GroupName, Version: "v1"}
+	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
+	// Install is a function which adds this version to a scheme
+	Install = SchemeBuilder.AddToScheme
+
+	// SchemeGroupVersion generated code relies on this name
+	// DEPRECATED
+	SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1"}
+	// AddToScheme exists solely to keep the old generators creating valid code
+	// DEPRECATED
+	AddToScheme = SchemeBuilder.AddToScheme
+)
+
+// Resource generated code relies on this being here, but it logically belongs to the group
+// DEPRECATED
+func Resource(resource string) schema.GroupResource {
+	return schema.GroupResource{Group: GroupName, Resource: resource}
+}
+
+func addKnownTypes(scheme *runtime.Scheme) error {
+	metav1.AddToGroupVersion(scheme, SchemeGroupVersion)
+
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&CredentialsRequest{}, &CredentialsRequestList{},
+		&AWSProviderStatus{}, &AWSProviderSpec{},
+		&AzureProviderStatus{}, &AzureProviderSpec{},
+		&GCPProviderStatus{}, &GCPProviderSpec{},
+		&VSphereProviderStatus{}, &VSphereProviderSpec{},
```

Comment:
This seems off. The usual k8s api is to have:
- spec - what user specified
- status - what the controller achieved

but in your case spec and status are separate, why? I believe this can't be changed at the moment but I must admit this is unexpected.

Reply:
This looks suspicious, all of these cloud specific types are embedded in the CredentialsRequest RawExtensions. I'm wondering if we need to add these as known types at all. It might be related to serialization as we do have to decode them in the cloud specific actuators.

Reply:
If these types are embedded I don't think they should be part of openshift/api at all.

Reply:
I actually considered removing this in the CCO repo (https://github.com/openshift/cloud-credential-operator/pull/219) as prep-work before opening this PR, but it was never merged. I can certainly resurrect that PR and stop registering these types.